pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,257 @@
+		}
+		defer results.Close()
+
+		for range m {
+			_, err := results.Exec()
+			if err != nil {
+				// Metadata table not accessible. Try for func in Promscale extension.
+				return fmt.Errorf("error querying results: %w", err)
+			}
+		}
+		return err
+	}
+
+	// Try to update via Promscale extension.
+	if err := update(metadataUpdateWithExtension); err != nil {
+		fmt.Println("err 1", err.Error())
+		// Promscale extension not installed. Try to attempt to write directly as a rare attempt
+		// in case we fix the _timescaledb_catalog.metadata permissions in the future.
+		_ = update(metadataUpdateNoExtension)
+	}
+
+	return nil
+}
+
+func (t *engine) Stop() {
+	t.stop <- struct{}{}
+}
+
+func (t *engine) Start() {
+	t.stop = make(chan struct{})
+	go func() {
+		collect := time.NewTicker(time.Minute * 55) // Collect telemetry info every 55 minutes since housekeeper fills the _timescaledb_catalog.metadata table every 1 hour.
+		defer collect.Stop()
+
+		for {
+			select {
+			case <-t.stop:
+				close(t.stop)
+				return
+			case <-collect.C:
+			}
+			var (
+				err             error
+				underlyingValue float64
+
+				newStats = make(map[string]float64)
+			)
+			t.metrics.Range(func(statName, metric interface{}) bool {
+				columnName := statName.(string)
+				promMetric := metric.(prometheus.Metric)
+				underlyingValue, err = extractMetricValue(promMetric)
+				if err != nil {
+					err = fmt.Errorf("extracting metric value of stat '%s': %w", columnName, err)
+					return false
+				}
+
+				newStats[columnName] = underlyingValue
+				return true
+			})
+			if err != nil {
+				log.Debug("msg", "error extracting underlying value of a stats metric", "error", err.Error())
+				continue
+			}
+			if err := syncInfoTable(t.conn, t.uuid, newStats); err != nil {
+				log.Debug("msg", "syncing new stats", "error", err.Error())
+			}
+		}
+	}()
+}
+
+func extractMetricValue(metric prometheus.Metric) (float64, error) {
+	var internal io_prometheus_client.Metric
+	if err := metric.Write(&internal); err != nil {
+		return 0, fmt.Errorf("error writing metric: %w", err)
+	}
+	if internal.Gauge != nil {
+		return internal.Gauge.GetValue(), nil
+	} else if internal.Counter != nil {
+		return internal.Counter.GetValue(), nil
+	}
+	return 0, fmt.Errorf("both Gauge and Counter are nil")
+}
+
+// RegisterMetric registers a counter or gauge metric for telemetry purpose.
+func (t *engine) RegisterMetric(columnName string, metric prometheus.Metric) error {
+	if !isCounterOrGauge(metric) {
+		return fmt.Errorf("metric not a counter or gauge")
+	}
+	t.metrics.Store(columnName, metric)
+	return nil
+}
+
+func isCounterOrGauge(metric prometheus.Metric) bool {
+	switch metric.(type) {
+	case prometheus.Counter, prometheus.Gauge:
+		return true
+	default:
+		return false
+	}
+}
+
+// syncInfoTable stats with promscale_instance_information table.
+func syncInfoTable(conn pgxconn.PgxConn, uuid [16]byte, stats map[string]float64) error {
```

Comment:
This needs to be tested in some tests.

Reply:
Also, why is this not a method on the engine? Seems to be using engine connection and uuid.